packages/loot-core/src/server/budget/categoryTemplate.ts

```diff
@@ -0,0 +1,539 @@
+      this.toBudgetAmount += toBudget + 1;
+    } else {
+      this.toBudgetAmount += toBudget;
+    }
+    return toBudget;
+  }
+
+  getValues() {
+    this.runGoal();
+    return { budgeted: this.toBudgetAmount, goal: this.goalAmount };
+  }
+
+  //-----------------------------------------------------------------------------
+  // Implimentation
+  readonly categoryID: string; //readonly so we can double check the category this is using
+  private month: string;
+  private templates = [];
+  private remainder = [];
+  private goals = [];
+  private priorities: number[] = [];
+  private remainderWeight: number = 0;
+  private toBudgetAmount: number = null; // amount that will be budgeted by the templates
+  private isLongGoal: boolean = null; //defaulting the goals to null so templates can be unset
+  private goalAmount: number = null;
+  private fromLastMonth = 0; // leftover from last month
+  private limitAmount = null;
+  private limitCheck = false;
+  private limitHold = false;
+
+  private constructor(
+    templates,
+    categoryID: string,
+    month: string,
+    fromLastMonth: number,
+  ) {
+    this.categoryID = categoryID;
+    this.month = month;
+    this.fromLastMonth = fromLastMonth;
+    // sort the template lines into regular template, goals, and remainder templates
+    if (templates) {
+      templates.forEach(t => {
+        if (t.directive === 'template' && t.type !== 'remainder') {
+          this.templates.push(t);
+        }
+      });
+      templates.forEach(t => {
+        if (t.directive === 'template' && t.type === 'remainder') {
+          this.remainder.push(t);
+        }
+      });
+      templates.forEach(t => {
+        if (t.directive === 'goal') this.goals.push(t);
+      });
+    }
+    // check limits here since it needs to save states inside the object
+    this.checkLimit();
+    this.checkSpend();
+    this.checkGoal();
+
+    //find priorities
+    const p = [];
+    this.templates.forEach(t => {
+      if (t.priority != null) {
+        p.push(t.priority);
+      }
+    });
+    //sort and reduce to unique items
+    this.priorities = p
+      .sort(function (a, b) {
+        return a - b;
+      })
+      .filter((item, idx, curr) => curr.indexOf(item) === idx);
+
+    //find remainder weight
+    let weight = 0;
+    this.remainder.forEach(r => {
+      weight += r.weight;
+    });
+    this.remainderWeight = weight;
+  }
+
+  private runGoal() {
+    if (this.goals.length > 0) {
+      this.isLongGoal = true;
+      this.goalAmount = amountToInteger(this.goals[0].amount);
+      return;
+    }
+    this.goalAmount = this.toBudgetAmount;
+  }
+
+  //-----------------------------------------------------------------------------
+  //  Template Validation
+  static async checkByAndScheduleAndSpend(templates, month) {
+    //check schedule names
+    const scheduleNames = (await getActiveSchedules()).map(({ name }) => name);
+    templates
+      .filter(t => t.type === 'schedule')
```

Comment:
If none of the templates are `schedule` this function should return before doing asynchronous actions.